525: update config file docs

NMontanaBrown opened the pull request:
# Description

Updating the config file documentation, as this was not very thorough considering all the possible fields and necessary configuration necessary to make it work. I think I've covered most of the config file fields, from what I can gather from examples and documentation as it stands for the code itself.

I'm looking for some feedback on the current content and style before we open the PR officially and add this to the RTD via merge.

Fixes #525

## Type of change

What types of changes does your code introduce to DeepReg?

_Please check the boxes that apply after submitting the pull request._

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [x] Documentation Update (fix or improvement on the documentation)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Other (if none of the other choices apply)

## Checklist

_Please check the boxes that apply after submitting the pull request._

_If you're unsure about any of them, don't hesitate to ask. We're here to help! This is
simply a reminder of what we are going to look for before merging your code._

- [x] I have
      [installed pre-commit](https://deepreg.readthedocs.io/en/latest/contributing/setup.html)
      using `pre-commit install` and formatted all changed files. If you are not
      certain, run `pre-commit run --all-files`.
- [x] My commits' message styles matches
      [our requested structure](https://deepreg.readthedocs.io/en/latest/contributing/commit.html),
      e.g. `Issue #<issue number>: detailed message`.
- [x] I have updated the
      [change log file](https://github.com/DeepRegNet/DeepReg/blob/main/CHANGELOG.md)
      regarding my changes.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

NMontanaBrown commented:
Hi @mathpluscode could you have a read over the current content of this PR? It's messy for style but I can work on that before merge. Looking to check veracity of what I've added to the docs in terms of config.

mathpluscode commented on docs/source/docs/configuration.md:
````diff
@@ -2,19 +2,459 @@
+    test: "data/test/h5/paired/test" # folder contains test data
+```
+
+Multiple dataset directories can be specified, such that data is sampled across several
````
data are plural 

https://www.theguardian.com/news/datablog/2010/jul/16/data-plural-singular

mathpluscode commented on docs/source/docs/configuration.md:
```diff
@@ -2,19 +2,459 @@
+The type of data we supply the data loaders will influence the behaviour, so we must
+specify the data type using the `format` key. Currently, DeepReg data loaders support
+nifti and h5 file types - alternate file formats will raise errors in the data loaders.
+To indicate which format to use, pass a string to this field as either "nifti" or "h5":
```
maybe specify that h5 data must end with `.h5` and nifti must end with `.nii` or `.nii.gz`

mathpluscode commented on docs/source/docs/configuration.md:
```diff
@@ -2,19 +2,459 @@
+      - "data/test/h5/paired/train1"
+      - "data/test/h5/paired/train2" # folders contains training data
+    valid: "data/test/h5/paired/valid" # folder contains validation data
+    test: "data/test/h5/paired/test" # folder contains test data
```
do you think we should add other required args in these previous examples as well?

NMontanaBrown replied:
So my idea was to give examples on "how to build it" bit by bit, so that the different parts don't get confused with each other. Is it worth adding a sentence about this to clarify?

mathpluscode commented on docs/source/docs/configuration.md:
```diff
@@ -2,19 +2,459 @@
+### Labeled key - Required
+
+The `labeled` key indicates whether labels should be used during training. A Boolean is
+used to indicate the usage of labels:
```
boolean values are case sensitive, we recommend using only `true` and `false`, all lower case

https://pyyaml.org/wiki/PyYAMLDocumentation

mathpluscode commented on docs/source/docs/configuration.md:
```diff
@@ -2,19 +2,459 @@
+### Backbone - required
+
+The `backbone` section defines the backbone network for the registration. This section
+has several subsections. The `name` and `num_channel_initial` are global to all backbone
```
only `name` is required, all other params will be passed to the backbone class directly, https://github.com/DeepRegNet/DeepReg/blob/ba27961deb0a3d1721747e03ae5a25ef49d988ea/deepreg/model/network/util.py#L56

maybe link the doc to some code can be helpful?

NMontanaBrown replied:
Yeah, so I was going off the implem prior to the registry. How would this work then? We still need to pass  those args into the config right?

NMontanaBrown replied:
@mathpluscode I am trying to resolve this now but I still don't understand how the config changes from the Registry code. So, do we have something like this?

```yaml
train:
  method: "ddf" # One of ddf, dvf, conditional
  name:
    backbone: "unet"  # One of unet, local, global
    num_channel_initial: 16 # Int type, number of initial channels in the network. Controls the network size.
    depth: 3
    pooling: False
    concat_skip: True
```

where do we put the class name? under name? I am not sure how to proceed at the moment.

NMontanaBrown replied:
@YipengHu  see this discussion

mathpluscode replied:
oh sorry I missed this msg, is it clear now for you?

mathpluscode replied:
it should be 
```
train:
  method: "ddf" # One of ddf, dvf, conditional
  backbone:
    name: "unet"  # One of unet, local, global
    num_channel_initial: 16 # Int type, number of initial channels in the network. Controls the network size.
    depth: 3
    pooling: False
    concat_skip: True
```

mathpluscode replied:
my bad, i should provide the example for u

NMontanaBrown replied:
Solved.

mathpluscode commented on docs/source/docs/configuration.md:
```diff
@@ -2,19 +2,459 @@
+#### Image
+
+- `weight`: float type, weight of individual loss element in total loss function.
+- `name`: string type, one of "lncc", "ssd" or "gmi". EG.
```
what's the EG in the end

mathpluscode commented on docs/source/docs/configuration.md:
```diff
@@ -2,19 +2,459 @@
+### Optimizer - required
+
+The optimizer can be defined by using a `name` and then passing optimizer specific
+arguments with the same name. All optimizers can use the `learning_rate` argument.
```
same as backbone, except the `name` all others will be passed to the class directly

mathpluscode replied:
oh.. seems like i haven't do this refactoring yet. forget this thread

mathpluscode commented:
@NMontanaBrown thx sooooo much for the efforts!!!!!

NMontanaBrown commented on docs/source/docs/configuration.md:
```diff
@@ -2,19 +2,459 @@
+  - `multi_scale`:
+    - `loss_type`: string type, name of loss to use. Options currently include "dice",
+      "cross-entropy", "mean-squared", "generalised_dice" and "jaccard".
+    - `loss_scales`: list of ints, what
```
Didn't finish my own sentence here...

mathpluscode commented:
@NMontanaBrown if this PR is ready, I suggest to merge it asap? I'm about to change losses to prepare for multi loss. see #567

NMontanaBrown commented:
> @NMontanaBrown if this PR is ready, I suggest to merge it asap? I'm about to change losses to prepare for multi loss. see #567

Hi @mathpluscode  I am currently out of office. I will not be working on this until next week earliest. Can you wait until then?

mathpluscode commented:
> > @NMontanaBrown if this PR is ready, I suggest to merge it asap? I'm about to change losses to prepare for multi loss. see #567
> 
> Hi @mathpluscode I am currently out of office. I will not be working on this until next week earliest. Can you wait until then?

I will continue my work, in case of changes of configs, later we can sync and I'm more than happy to correct the corresponding configs on your branch :)

mathpluscode commented:
@NMontanaBrown I propose to have a meeting/discussion regarding this branch :)

NMontanaBrown commented:
Just run over the Registry and new updates with @mathpluscode, resolution is that he will finish the updates on registry, request review, then we merge here and finish docs (only a few changes as nominally the args have different names).

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/DeepRegNet/DeepReg/inspections/d45a535a-45a0-4f39-b211-9a775226abc4) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/DeepRegNet/DeepReg/inspections/d45a535a-45a0-4f39-b211-9a775226abc4) completed: **No new issues**

mathpluscode commented:
@NMontanaBrown Let me know when this is ready?

And we can add a comparison between last released version config and the current config ^^
To detail what are the changes exactly.

NMontanaBrown commented:
> @NMontanaBrown Let me know when this is ready?
> 
> And we can add a comparison between last released version config and the current config ^^
> To detail what are the changes exactly.

What do you mean "add a comparison between last released version config and current config"?

YipengHu commented:
> > @NMontanaBrown Let me know when this is ready?
> > And we can add a comparison between last released version config and the current config ^^
> > To detail what are the changes exactly.
> 
> What do you mean "add a comparison between last released version config and current config"?

a side-by-side diff

mathpluscode commented:
> > > @NMontanaBrown Let me know when this is ready?
> > > And we can add a comparison between last released version config and the current config ^^
> > > To detail what are the changes exactly.
> > 
> > 
> > What do you mean "add a comparison between last released version config and current config"?
> 
> a side-by-side diff?

@NMontanaBrown @YipengHu 

Yes, a separate section, showing one config before and one config afterwards, to highlight the difference ^^
So that user can adapt their config files easily ^^

NMontanaBrown commented:
@mathpluscode @YipengHu Before we continue any further can we confirm #627 and #624? Doesn't make sense to merge this if there are still potential errors in config, at least for #624.

mathpluscode commented:
> @mathpluscode @YipengHu Before we continue any further can we confirm #627 and #624? Doesn't make sense to merge this if there are still potential errors in config, at least for #624.

@NMontanaBrown go ahead, it is just a typo in the code, no worry regarding the config docs.
Also, I suggest you merge this branch asap ^^ so that for my model refactoring, I can handle the config doc changes

mathpluscode commented on docs/source/docs/configuration.md:
```diff
@@ -2,19 +2,515 @@
+    num_channel_initial: 16 # Int type, number of initial channels in the network. Controls the network size.
+    extract_levels: [0, 1, 2]
+  loss:
+    dissimilarity:
```
in the latest `main` this `dissimilarity` layer has been removed.

@NMontanaBrown if you want I can update this branch for you

NMontanaBrown replied:
The only thing left is the multi-scales argument. So now, how do we do scales for losses? previously we had:

```yaml
loss:
    image:
      name: "gmi"
      weight: 1.0
    label:
      weight: 1.0
      name: "multi_scale" # "multi_scale" or "single_scale"
      multi_scale:
        loss_type: "dice" # options include "dice", "cross-entropy", "mean-squared", "generalised_dice" and "jaccard"
        loss_scales: [0, 1, 2, 4, 8, 16, 32]
```

now.... we don't specify the argument "multi-scale" or single scale, we just pass the name: "dice", then scales: a list, and if list len >1, it is understood as multiscale?

NMontanaBrown replied:
so it would be:
```yaml
label:
      weight: 1.0
      name: "dice"
      scales: [0, 1, 2, 4, 8, 16, 32]
```
is interpreted as multi-scale dice
and then : 
```yaml
label:
      weight: 1.0
      name: "dice"
      scales: [1]
```
is interpreted as single scale dice?

mathpluscode replied:
check https://github.com/DeepRegNet/DeepReg/blob/main/deepreg/model/loss/label.py#L42

So if you want to use single scale` just not to put `scales` at all, or leave empty, [1] works too

is it understandable? any suggestions?

mathpluscode replied:
let me be more detailed, to be single scale, we can do

```yaml
label:
      weight: 1.0
      name: "dice"
      scales: [0]
```
important point, no scaling means value = 0

or

```yaml
label:
      weight: 1.0
      name: "dice"
      scales: null
```

or

```yaml
label:
      weight: 1.0
      name: "dice"
      scales: 
```

But we should not give an empty list (for now)

```yaml
label:
      weight: 1.0
      name: "dice"
      scales: []
```

mathpluscode replied:
maybe I can update the code, to make `[]` -> single scale ^^, what do you think

NMontanaBrown replied:
Updating, I just saw this. Add ticket for empty list to be single scale. I will include in the docs at the moment that that will break the code.

NMontanaBrown commented:
From label code, what does [neg_weight = weight for negative class](https://github.com/DeepRegNet/DeepReg/blob/859df2a91564d16c43ea761ad272de568b8784bb/deepreg/model/loss/label.py#L224) mean? Do you mean for 0 (background) class?

mathpluscode commented:
> From label code, what does [neg_weight = weight for negative class](https://github.com/DeepRegNet/DeepReg/blob/859df2a91564d16c43ea761ad272de568b8784bb/deepreg/model/loss/label.py#L224) mean? Do you mean for 0 (background) class?

Ah yes, sorry for the naming, do you think it's better to renamed to `background_weight` ?

NMontanaBrown commented:
Raise a ticket for this, it's minor but I'm not going to mix it up here.

mathpluscode commented:
> Raise a ticket for this, it's minor but I'm not going to mix it up here.

raised in #632, will fix that after this branch

NMontanaBrown commented:
[Jaccard doesn't have a neg_weight argument in get_config in comparison to the other losses. Is there a reason for this?](https://github.com/DeepRegNet/DeepReg/blob/859df2a91564d16c43ea761ad272de568b8784bb/deepreg/model/loss/label.py#L319)

YipengHu commented:
> [Jaccard doesn't have a neg_weight argument in get_config in comparison to the other losses. Is there a reason for this?](https://github.com/DeepRegNet/DeepReg/blob/859df2a91564d16c43ea761ad272de568b8784bb/deepreg/model/loss/label.py#L319)

not implemented. a ticket maybe

mathpluscode commented:
> [Jaccard doesn't have a neg_weight argument in get_config in comparison to the other losses. Is there a reason for this?](https://github.com/DeepRegNet/DeepReg/blob/859df2a91564d16c43ea761ad272de568b8784bb/deepreg/model/loss/label.py#L319)

no, we can add that for sure, new issue again then XD

mathpluscode commented:
> > [Jaccard doesn't have a neg_weight argument in get_config in comparison to the other losses. Is there a reason for this?](https://github.com/DeepRegNet/DeepReg/blob/859df2a91564d16c43ea761ad272de568b8784bb/deepreg/model/loss/label.py#L319)
> 
> no, we can add that for sure, new issue again then XD

I updated the issue description https://github.com/DeepRegNet/DeepReg/issues/632, so we handle the `background_weight` all together!

NMontanaBrown commented:
Also, previously the following losses were supported but now we have deleted them from our loss repertoire, as I cannot find them in the new label loss dict? And we don't add them via add_loss in model right? :

`mean-squared`
`generalised_dice`

mathpluscode commented:
> Also, previously the following losses were supported but now we have deleted them from our loss repertoire, as I cannot find them in the new label loss dict? And we don't add them via add_loss in model right? :
> 
> `mean-squared`
> `generalised_dice`

`mead-squared` is the same as SSD, `generalised_dice` has been merged into `Dice` :)

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/DeepRegNet/DeepReg/inspections/c043215c-9586-45f1-9878-bbd6a59fa068) was created.

NMontanaBrown commented:
> > Also, previously the following losses were supported but now we have deleted them from our loss repertoire, as I cannot find them in the new label loss dict? And we don't add them via add_loss in model right? :
> > `mean-squared`
> > `generalised_dice`
> 
> `mead-squared` is the same as SSD, `generalised_dice` has been merged into `Dice` :) @NMontanaBrown

But strictly speaking, ssd loss is called on the image right? So we have removed ssd/mean_squared from label losses?

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/DeepRegNet/DeepReg/inspections/c0e53815-00ac-47d7-a936-5b0ef4fc06c8) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/DeepRegNet/DeepReg/inspections/c043215c-9586-45f1-9878-bbd6a59fa068) completed: **No new issues**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/DeepRegNet/DeepReg/inspections/0f3d4545-5075-4d1e-b79e-0a66cf6bc188) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/DeepRegNet/DeepReg/inspections/aa5b0c65-523f-4cda-a785-93fc43dfe001) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/DeepRegNet/DeepReg/inspections/c0e53815-00ac-47d7-a936-5b0ef4fc06c8) completed: **No new issues**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/DeepRegNet/DeepReg/inspections/6117caf1-823f-48b1-9299-b93e82d496ed) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/DeepRegNet/DeepReg/inspections/aa5b0c65-523f-4cda-a785-93fc43dfe001) completed: **No new issues**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/DeepRegNet/DeepReg/inspections/6117caf1-823f-48b1-9299-b93e82d496ed) completed: **No new issues**

NMontanaBrown commented:
I would personally prefer > 2 reviewers on this MR, if possible.

mathpluscode commented on docs/source/docs/configuration.md:
````diff
@@ -2,19 +2,592 @@
+```yaml
+dataset:
+  dir:
+    train:
````
the `folders contains training data` should be in line 34?
btw shouldn't it be `containing`?

mathpluscode commented on docs/source/docs/configuration.md:
```diff
@@ -2,19 +2,592 @@
+
+### Format key - Required
+
+The type of data we supply the data loaders will influence the behaviour, so we must
```
I would say `The data file format`? as for type, we may think about labeled, or unlabeled? also to be consistent with the key `format`

mathpluscode commented on docs/source/docs/configuration.md:
````diff
@@ -2,19 +2,592 @@
+  format: "nifti"
+```
+
+Depending on the data type, DeepReg will expect the images and labels to be stored in
````
Deepreg expects ... (i personally find this better)

mathpluscode commented on docs/source/docs/configuration.md:
```diff
@@ -2,19 +2,592 @@
+### Type key - Required
+
+The type of data loader used will depend on how one wants to train the network.
+Currently, DeepReg data loaders support the "paired", "unpaired" and "grouped" training
```
do you want to write `paired` instead of "paired"? it's minor for me though

mathpluscode commented on docs/source/docs/configuration.md:
```diff
@@ -2,19 +2,592 @@
+
+###### Sample_label - Required
+
+In the case that we have more than one label per image, we need to indicate to the
```
"inform the loader" instead of "indicate to the loader"?

mathpluscode commented on docs/source/docs/configuration.md:
````diff
@@ -2,19 +2,592 @@
+```yaml
+dataset:
+  dir:
+    train: "data/test/h5/paired/train1" # folders contains training data
````
typo? the `train1` you got 1 here

Btw, it is one folder

NMontanaBrown replied:
Oh, this was on purpose to specifically highlight using just one folder.

mathpluscode commented on docs/source/docs/configuration.md:
````diff
@@ -2,19 +2,592 @@
+  sample_label: "sample" # one of "sample", "all" or None
+```
+
+In the case the `labeled` argument is false, the sample_label is unused, but still must
````
oh, is this true, then we should optimize this part. I will do a new issue.

mathpluscode commented on docs/source/docs/configuration.md:
````diff
@@ -2,19 +2,592 @@
+```
+
+In the case the `labeled` argument is false, the sample_label is unused, but still must
+be passed. Additionally, if the tensors in the files only have one label, irregardles of
````
regardless or irregardless ?

NMontanaBrown replied:
you are correct!

mathpluscode commented on docs/source/docs/configuration.md:
```diff
@@ -2,19 +2,592 @@
+
+##### Paired
+
+- `moving_image_shape`: (list, tuple) len 3, corresponding to (dim1, dim2, dim3) of the
```
use a list or a tuple of three ints? more understandable ^^

mathpluscode replied:
btw, these are from the docstring? maybe better to link to the function? as if we change the docstring, we have to change here as well (which is hard to detect/remember)

NMontanaBrown replied:
If we change the functions from using list/tuples, we will need to modify the configs again, as the new configs will not be backcompatible in that case. So, maybe we should maintain the configs and for new changes away from the established in this config for 1.0.0 may need intermediate functions to adapt to this config? I can still link but these are my thoughts

mathpluscode replied:
oh, i mean not for this single one only.

As here it seems that you copied the description of the args from multiple functions. I guess a relative link to the code would be easier to maintain?

mathpluscode commented:
thx @NMontanaBrown I added my comments ^^ otherwise all good

YipengHu commented on docs/source/docs/configuration.md:
```diff
@@ -2,19 +2,592 @@
+In the case the `labeled` argument is false, the sample_label is unused, but still must
+be passed. Additionally, if the tensors in the files only have one label, irregardles of
+the `sample_label` argument, the data loader will only pass the one label to the
+network.
```
it is not clear, we have a more detailed explanation to this cite the link: https://deepreg.readthedocs.io/en/latest/docs/exp_label_sampling.html

also, @mathpluscode what is the difference between "all" and "sample" during training? - it is not clear.

mathpluscode replied:
For training,
- `all` means for each image we will yield image-label pairs for all labels
- `sample` means for each image we will yield only one image-label pair

YipengHu replied:
how can you yield all pairs during training?

YipengHu replied:
random with replacement

NMontanaBrown replied:
Ok, to clarify:
- all: for one image, x labels: yelds x image label pairs with the same image a
- sample: for one image, with x labels: yields 1 image label pair randomly sampled.

YipengHu replied:
@mathpluscode please clarify/correct this. If it is still sample one at a time, we should separately explain this for train and test, as you can't use more than one image-label pair at a time:
During training:
all: sampling without replacement
sample: sampling with replacement
During testing:
all: sampling without replacement until all labels exhausted
sample: randomly sampling one image-label pair at a test time

NMontanaBrown replied:
@YipengHu @mathpluscode

YipengHu replied:
@NMontanaBrown - had a chat with @mathpluscode I think we can clarify this by:
"sample_label" specifies the label sampling method to construct an epoch of image-label pairs for training. 
The default for training is the "sample" option, which randomly samples one label for each image to form an epoch, resulting one image-label pair for each available image. The option "all" samples all available labels for each image, resulting in larger and unequal number of image-label pairs. The "all" is the default option during test stage predicting for all possible image-label pairs.

NMontanaBrown replied:
So, then...
Ok, to clarify:
- all: for one image, x labels: yields x image label pairs with the same image a. Occurs over all images, in one epoch.
- sample: for one image, with x labels: yields 1 image label pair randomly sampled. Occurs for all images in one epoch?

YipengHu commented on docs/source/docs/configuration.md:
```diff
@@ -2,19 +2,592 @@
+  group, so one epoch has num_groups pairs of data, if false, iterate through this
+  loader will generate all possible pairs.
+- `image_shape`: (list, tuple) len 3, corresponding to (dim1, dim2, dim3) of the 3D
+  image.
```
again worth citing https://deepreg.readthedocs.io/en/latest/docs/exp_label_sampling.html

YipengHu commented on docs/source/docs/configuration.md:
```diff
@@ -2,19 +2,592 @@
+arguments under the same indent. The first argument should be the argument `name`, which
+should be string type, one of "unet", "local" or "global", to define a UNet, LocalNet or
+GlobalNet backbone, respectively. With Registry functionalities, you can also define
+your own networks to pass to DeepReg train via config.
```
this needs a placeholder for link to registry

mathpluscode replied:
the registry is not only for backbone @YipengHu, we should link on the top of this doc

https://deepreg.readthedocs.io/en/latest/docs/registry.html

btw, better to use relative link instead of URL link :)

YipengHu commented on docs/source/docs/configuration.md:
```diff
@@ -2,19 +2,592 @@
+  backbone:
+    name: "local" # One of unet, local, global
+    num_channel_initial: 16 # Int type, number of initial channels in the network. Controls the network size.
+    extract_levels: [0, 1, 2]
```
"global" net does not support extract_levels - worth adding a comment after this line

NMontanaBrown replied:
Are you sure? https://github.com/DeepRegNet/DeepReg/blob/859df2a91564d16c43ea761ad272de568b8784bb/deepreg/model/backbone/global_net.py#L30

NMontanaBrown replied:
@YipengHu

YipengHu replied:
> Are you sure?
> 
> https://github.com/DeepRegNet/DeepReg/blob/859df2a91564d16c43ea761ad272de568b8784bb/deepreg/model/backbone/global_net.py#L30

yup - the global net predicts parameters of transformation models, such as affine, without decoder. so this is redundant @mathpluscode  a ticket please

NMontanaBrown replied:
Ok regardless of the fact of it being redundant or not... What to do with the docs? @YipengHu @mathpluscode

YipengHu replied:
good question. for this PR, we can only say "this is not used for global net, will be deprecated in future release" (and remember to do it...)

NMontanaBrown replied:
Tickeeeet

YipengHu commented on docs/source/docs/configuration.md:
```diff
@@ -2,19 +2,592 @@
+#### Local and GlobalNet
+
+The LocalNet has an encoder-decoder structure, and extracts information from tensors at
+certain levels. We can define which levels to extract info from with the
```
"certain levels" -> one or multiple resolution levels

YipengHu commented on docs/source/docs/configuration.md:
```diff
@@ -2,19 +2,592 @@
+and regularization losses (on the DDFs predicted).
+
+DeepReg uses `tf.keras.Model` `add_loss()` in the Registry method to add losses to the
+model, which provides some flexibility in configuration.
```
this is implementation details - should be in docstring, not here

NMontanaBrown replied:
Actually don't agree - given there are no docs on registry yet, this information is useful for users if they want to add new losses to their models.

YipengHu commented on docs/source/docs/configuration.md:
```diff
@@ -2,19 +2,592 @@
+This section defines the loss in training.
+
+The losses in DeepReg are defined depending on the type of network to be built, and can
+be split into three sections: imag and label losses (between moving and fixed tensors),
```
typo: imag

YipengHu commented on docs/source/docs/configuration.md:
```diff
@@ -2,19 +2,592 @@
+Currently, DeepReg offers conditional, ddf/dvf and affine registration pre-built models.
+Traditionally, models have been configured with the following losses:
+
+- Conditional: label loss.
```
@mathpluscode does conditional not support image loss?

mathpluscode replied:
no, as you consume both images as input

NMontanaBrown replied:
No, see [L68](https://github.com/DeepRegNet/DeepReg/blob/859df2a91564d16c43ea761ad272de568b8784bb/deepreg/model/network/cond.py#L68)

YipengHu commented on docs/source/docs/configuration.md:
```diff
@@ -2,19 +2,592 @@
+
+The above sections are necessary to build a model correctly. Not passing all sections as
+defined above may raise errors. Currently you can call one loss per field eg. one label
+loss type, one image loss type and one ddf/dvf loss type.
```
also: you can set weight to 0 to not use it. weights on label loss will be ignored if the "labeled" key was false or segmentation labels are not available.

YipengHu commented on docs/source/docs/configuration.md:
```diff
@@ -2,19 +2,592 @@
+
+#### Image
+
+The image loss calculates dissimilarity between image tensors passed through the
```
"image tensors passed though the network" -> "warped image tensors and fixed image tensors"
n.b. network predicts ddfs to warp images, so strictly speaking, no images are "through" the network...

YipengHu commented on docs/source/docs/configuration.md:
````diff
@@ -2,19 +2,592 @@
+      weight: 0.1
+```
+
+The following are the default image losses. Additional arguments should be added at the
````
is it default image losses or "default parameters for each image loss"?

@mathpluscode do we have a "default" loss?

mathpluscode replied:
there is no default loss, you have to define one, each loss may have some default args

NMontanaBrown replied:
1) Default image losses: default implemented deepreg losses.
2) default loss: certain loss configs are ascribed depending on model method.

NMontanaBrown replied:
How about: The following are the DeepReg image losses @YipengHu

YipengHu commented on docs/source/docs/configuration.md:
```diff
@@ -2,19 +2,592 @@
+- `gmi`: Calls a global mutual informatin loss. Requires the following arguments:
+  - `num_bins`: int, optional, default=23. Number of bins for intensity.
+  - `sigma_ratio`: float, optional, default=0.5. A hyper parameter for the gaussian
+    function.
```
"A hyper parameter for the gaussian function." - > "A hyperparameter for the Gaussian kernel density estimation"

YipengHu commented on docs/source/docs/configuration.md:
```diff
@@ -2,19 +2,592 @@
+
+The label loss calculates dissimilarity between labels.
+
+All pre-implemented losses can be used as multi-scale or single scale losses.
```
what is the difference between "pre-implemented" and "implemented" here?

NMontanaBrown replied:
Changing to default.

YipengHu commented on docs/source/docs/configuration.md:
```diff
@@ -2,19 +2,592 @@
+- `weight`: float type, weight of individual loss element in total loss function.
+- `scales`: list, or None. Optional argument. If you do not pass this argument (or pass
+  the list [0], the value `null` or an empty value pair), the loss is calculated at a
+  single scale. If you past a list of length > 1, a multi-scale loss will be used.
```
typo: past

YipengHu commented on docs/source/docs/configuration.md:
```diff
@@ -2,19 +2,592 @@
+
+- `dice`: Calls a Dice loss on the labels, requires the following arguments:
+
+  - `binary`: bool, default is false. If true, the tensors are thresholded at 0.5.
```
does it have gradient when true? @mathpluscode if not, we should add the warning here

mathpluscode replied:
i think yes, like relu, it still yields gradients

YipengHu commented on docs/source/docs/configuration.md:
```diff
@@ -2,19 +2,592 @@
+
+  - `binary`: bool, default is false. If true, the tensors are thresholded at 0.5.
+  - `neg_weight`: float, default=0.0. Defines the weight of the background class (0)
+    versus foreground class, which is weighted as (1-neg_weight).
```
@mathpluscode please clarify this, default weight should be "no extra weighting between foreground and background classes", what is the range?

mathpluscode replied:
what do you mean "no extra weighting between foreground and background classes"

range should be [0, 1]

so, as you see here these stuff should be in docstring, and we only update docstring

YipengHu replied:
for Dice, default can be 1 for foreground - the alternative is to have both foreground and background in a generalised form

mathpluscode replied:
dice implementation here: https://github.com/DeepRegNet/DeepReg/blob/main/deepreg/model/loss/label.py#L123

this `neg_weight` = 0 gives normal dice,

anyway the whole idea is
- 0 means normal loss
- 1 means revert foreground background

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/DeepRegNet/DeepReg/inspections/4d98ce7f-cb0e-49bc-a08b-1ddc62395c4b) was created.

YipengHu commented on docs/source/docs/configuration.md:
```diff
@@ -2,19 +2,592 @@
+  arguments:
+
+  - `binary`: bool, default is false. If true, the tensors are thresholded at 0.5.
+  - `neg_weight`: float, default=0.0. Defines the weight of the background class (0)
```
the same here, traditionally, positive weight by default should be 1, while you can reduce it between (0,1] and increase it >1, such that pos_weight / 1 is the ratio between positive and negative classes, not pos_weight / (1-pos_weight). - similar for neg_weight - if somehow normalised, it should be 0.5 / 0.5 as default and otherwise should be treated as a bug @mathpluscode

mathpluscode replied:
emmmm maybe it's the name misleading,

the neg_weight = 0, means we treat 1 as foreground, 0 as background to calculate the loss

but we can treat 0 as foreground and 1 as background to calculate another loss, neg_weight is the weight for this loss

mathpluscode replied:
all this stuff comes from the generalised_dice :(

mathpluscode replied:
for the redundant between doc and docstring, I will do a separate issue

YipengHu replied:
ok, makes sense - only thing we need to change is here @NMontanaBrown to something like "positive and negative classes will be treated as (1-neg_weight) and neg_weight, respectively", for now, improve this later.

NMontanaBrown replied:
How about:
Defines the weight of the ngative (0)
    and positive (1) classes: by default, L = neg_weight*negative + (1-neg_weight)*positive

YipengHu replied:
i don't think it is correct @NMontanaBrown - as @mathpluscode said, this is really not weights between foreground and background losses, they are used to replace the 0 and 1 when calculating any loss (weights in a way i guess)

YipengHu replied:
you can also be explicit saying "neg_weight weights the foreground and background classes by replacing the labels of 1s and 0s with (1-neg_weight) and neg_weight, respectively"

NMontanaBrown replied:
Ok thanks, that is much clearer :)

mathpluscode commented:
@NMontanaBrown I understand that you would like to have more people to review.

Can I propose to merge this quickly, as we are modifying configs now and later, we will change this config many more times, and we will review it multiple times again and again, so in case of small issues, we can always fix them ^^ What do you think

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/DeepRegNet/DeepReg/inspections/4d98ce7f-cb0e-49bc-a08b-1ddc62395c4b) completed: **No new issues**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/DeepRegNet/DeepReg/inspections/9fb137ae-98e1-400b-84bb-82240f677eb6) was created.

YipengHu commented on docs/source/docs/configuration.md:
```diff
@@ -2,19 +2,592 @@
+- `name`: string type, the type of deformation energy to compute. Options include
+  "bending", "gradient"
+
+If the `gradient` loss is used, another optional argument can be passed at the same
```
why "another"?

NMontanaBrown replied:
It's not optional, actually. Since get_config will return error as it calls "l1" from the dict. Nice catch.

YipengHu commented on docs/source/docs/configuration.md:
```diff
@@ -2,19 +2,592 @@
+arguments with the same name. All optimizers can use the `learning_rate` argument.
+
+- `name`: string type, is used to define the optimizer during training. One of "adam",
+  "sgd", "rms". There must be another additional field with the same name.
```
not clear: "another additional field with the same name"?

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/DeepRegNet/DeepReg/inspections/9fb137ae-98e1-400b-84bb-82240f677eb6) completed: **No new issues**

YipengHu commented on docs/source/docs/configuration.md:
```diff
@@ -2,19 +2,592 @@
+  can be empty, which initalises a default
+  [Keras SGD optimizer](https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/SGD).
+  Alternatively, fields with names equivalent to those specified in the optimizer
+  documentation can be used instead.
```
"can be used" - > "can be configured by adding additional field with the same name" here?

YipengHu replied:
with the following examples, it is clear maybe...

NMontanaBrown commented:
> @NMontanaBrown I understand that you would like to have more people to review.
> 
> Can I propose to merge this quickly, as we are modifying configs now and later, we will change this config many more times, and we will review it multiple times again and again, so in case of small issues, we can always fix them ^^ What do you think

Waiting for review from another person etc

mathpluscode commented on docs/source/docs/configuration.md:
```diff
@@ -2,19 +2,592 @@
-two sections, `dataset` and `train`.
+two sections, `dataset` and `train`. Within `dataset` one specifies the data types,
+sizes, as well as the data loader to use. The `train` section specifies parameters
+related to the neural network. `yaml` files are used as they can be easily generated
```
agree with @YipengHu

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/DeepRegNet/DeepReg/inspections/c296a9ab-7151-4081-af02-5b8960624b90) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/DeepRegNet/DeepReg/inspections/c296a9ab-7151-4081-af02-5b8960624b90) completed: **No new issues**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/DeepRegNet/DeepReg/inspections/0fe986be-9d26-4f5f-a917-14671450fdfe) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/DeepRegNet/DeepReg/inspections/0fe986be-9d26-4f5f-a917-14671450fdfe) completed: **No new issues**

NMontanaBrown commented:
I am creating some graphics to summarise the config, what do you think: 
[DeepRegDatasetIllustration.pdf](https://github.com/DeepRegNet/DeepReg/files/5825820/DeepRegDatasetIllustration.pdf)
@mathpluscode @YipengHu

mathpluscode commented:
> I am creating some graphics to summarise the config, what do you think:
> [DeepRegDatasetIllustration.pdf](https://github.com/DeepRegNet/DeepReg/files/5825820/DeepRegDatasetIllustration.pdf)
> @mathpluscode @YipengHu

I wonder if this will be difficult to maintain.

YipengHu commented:
> > I am creating some graphics to summarise the config, what do you think:
> > [DeepRegDatasetIllustration.pdf](https://github.com/DeepRegNet/DeepReg/files/5825820/DeepRegDatasetIllustration.pdf)
> > @mathpluscode @YipengHu
> 
> I wonder if this will be difficult to maintain.
> Personally, a text version will be easier to maintain, and no need to consider the style/color/dark mode etc.

yeah thought it is clearer, it is difficult to maintain/adapt later, better to use any acceptable md/tex format?

NMontanaBrown commented:
These are fair comments. I think it is possible to do without markdown. I will open and work on this in another ticket. But I am still waiting on the sampling comment to be resolved @YipengHu @mathpluscode

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/DeepRegNet/DeepReg/inspections/5276af70-fb06-480c-90ed-a33da5908a02) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/DeepRegNet/DeepReg/inspections/5276af70-fb06-480c-90ed-a33da5908a02) completed: **No new issues**

YipengHu commented:
This is fine only thing is that for testing there is no such thing as an epoch.

Yipeng
________________________________
From: Nina Montana-Brown <notifications@github.com>
Sent: Sunday, January 17, 2021 3:51:28 PM
To: DeepRegNet/DeepReg <DeepReg@noreply.github.com>
Cc: Hu, Yipeng <yipeng.hu@ucl.ac.uk>; Mention <mention@noreply.github.com>
Subject: Re: [DeepRegNet/DeepReg] 525: update config file docs (#559)


@NMontanaBrown commented on this pull request.

________________________________

In docs/source/docs/configuration.md<https://github.com/DeepRegNet/DeepReg/pull/559#discussion_r559199902>:

> +```yaml

+dataset:

+  dir:

+    train: "data/test/h5/paired/train1" # folders contains training data

+    valid: "data/test/h5/paired/valid" # folder contains validation data

+    test: "data/test/h5/paired/test" # folder contains test data

+  format: "nifti"

+  type: "paired" # one of "paired", "unpaired" or "grouped"

+  labeled: true

+  sample_label: "sample" # one of "sample", "all" or None

+```

+

+In the case the `labeled` argument is false, the sample_label is unused, but still must

+be passed. Additionally, if the tensors in the files only have one label, irregardles of

+the `sample_label` argument, the data loader will only pass the one label to the

+network.



So, then...
Ok, to clarify:

  *   all: for one image, x labels: yields x image label pairs with the same image a. Occurs over all images, in one epoch.
  *   sample: for one image, with x labels: yields 1 image label pair randomly sampled. Occurs for all images in one epoch?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/DeepRegNet/DeepReg/pull/559#discussion_r559199902>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AATX37OZWLEOLVOIYNVY6JDS2MBQBANCNFSM4UWN4EFA>.

NMontanaBrown commented:
> This is fine only thing is that for testing there is no such thing as an epoch. Yipeng
> ________________________________ From: Nina Montana-Brown <notifications@github.com> Sent: Sunday, January 17, 2021 3:51:28 PM To: DeepRegNet/DeepReg <DeepReg@noreply.github.com> Cc: Hu, Yipeng <yipeng.hu@ucl.ac.uk>; Mention <mention@noreply.github.com> Subject: Re: [DeepRegNet/DeepReg] 525: update config file docs (#559) @NMontanaBrown commented on this pull request.
> ________________________________ In docs/source/docs/configuration.md<[#559 (comment)](https://github.com/DeepRegNet/DeepReg/pull/559#discussion_r559199902)>:
> +```yaml
> +dataset: + dir: + train: "data/test/h5/paired/train1" # folders contains training data + valid: "data/test/h5/paired/valid" # folder contains validation data + test: "data/test/h5/paired/test" # folder contains test data + format: "nifti" + type: "paired" # one of "paired", "unpaired" or "grouped" + labeled: true + sample_label: "sample" # one of "sample", "all" or None +``` + +In the case the `labeled` argument is false, the sample_label is unused, but still must +be passed. Additionally, if the tensors in the files only have one label, irregardles of +the `sample_label` argument, the data loader will only pass the one label to the +network. So, then... Ok, to clarify: * all: for one image, x labels: yields x image label pairs with the same image a. Occurs over all images, in one epoch. * sample: for one image, with x labels: yields 1 image label pair randomly sampled. Occurs for all images in one epoch? — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub<[#559 (comment)](https://github.com/DeepRegNet/DeepReg/pull/559#discussion_r559199902)>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AATX37OZWLEOLVOIYNVY6JDS2MBQBANCNFSM4UWN4EFA>.

But tests are not related to this docs ticket? @YipengHu

YipengHu commented:
> > This is fine only thing is that for testing there is no such thing as an epoch. Yipeng
> > ________________________________ From: Nina Montana-Brown [notifications@github.com](mailto:notifications@github.com) Sent: Sunday, January 17, 2021 3:51:28 PM To: DeepRegNet/DeepReg [DeepReg@noreply.github.com](mailto:DeepReg@noreply.github.com) Cc: Hu, Yipeng [yipeng.hu@ucl.ac.uk](mailto:yipeng.hu@ucl.ac.uk); Mention [mention@noreply.github.com](mailto:mention@noreply.github.com) Subject: Re: [DeepRegNet/DeepReg] 525: update config file docs (#559) @NMontanaBrown commented on this pull request.
> > ________________________________ In docs/source/docs/configuration.md<[#559 (comment)](https://github.com/DeepRegNet/DeepReg/pull/559#discussion_r559199902)>:
> > +`yaml +dataset: + dir: + train: "data/test/h5/paired/train1" # folders contains training data + valid: "data/test/h5/paired/valid" # folder contains validation data + test: "data/test/h5/paired/test" # folder contains test data + format: "nifti" + type: "paired" # one of "paired", "unpaired" or "grouped" + labeled: true + sample_label: "sample" # one of "sample", "all" or None +` + +In the case the `labeled` argument is false, the sample_label is unused, but still must +be passed. Additionally, if the tensors in the files only have one label, irregardles of +the `sample_label` argument, the data loader will only pass the one label to the +network. So, then... Ok, to clarify: * all: for one image, x labels: yields x image label pairs with the same image a. Occurs over all images, in one epoch. * sample: for one image, with x labels: yields 1 image label pair randomly sampled. Occurs for all images in one epoch? — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub<[#559 (comment)](https://github.com/DeepRegNet/DeepReg/pull/559#discussion_r559199902)>, or unsubscribehttps://github.com/notifications/unsubscribe-auth/AATX37OZWLEOLVOIYNVY6JDS2MBQBANCNFSM4UWN4EFA.
> 
> But tests are not related to this docs ticket? @YipengHu

why not - the config files used for both train and predict - see #634. @mathpluscode please clarify this. Is it true if you want different label_sample during train and predict, then you need to change this field "manually". you can't specify them differently in one config files - this can be addressed later but needs to clarify here.

YipengHu commented on docs/source/docs/configuration.md:
```diff
@@ -2,19 +2,609 @@
+the `extract_levels` argument.
+
+The GlobalNet encodes the image and uses the bottleneck layer to output an affine
+transformation using a FCN.
```
strictly speaking, this is not FCN as it has a fully-connected layer in the end for predicting the parameters. Consider "CNN"

YipengHu replied:
last bit - sorry being picky. otherwise looks good to merge.

mathpluscode commented:
> > > This is fine only thing is that for testing there is no such thing as an epoch. Yipeng
> > > ________________________________ From: Nina Montana-Brown [notifications@github.com](mailto:notifications@github.com) Sent: Sunday, January 17, 2021 3:51:28 PM To: DeepRegNet/DeepReg [DeepReg@noreply.github.com](mailto:DeepReg@noreply.github.com) Cc: Hu, Yipeng [yipeng.hu@ucl.ac.uk](mailto:yipeng.hu@ucl.ac.uk); Mention [mention@noreply.github.com](mailto:mention@noreply.github.com) Subject: Re: [DeepRegNet/DeepReg] 525: update config file docs (#559) @NMontanaBrown commented on this pull request.
> > > ________________________________ In docs/source/docs/configuration.md<[#559 (comment)](https://github.com/DeepRegNet/DeepReg/pull/559#discussion_r559199902)>:
> > > +`yaml +dataset: + dir: + train: "data/test/h5/paired/train1" # folders contains training data + valid: "data/test/h5/paired/valid" # folder contains validation data + test: "data/test/h5/paired/test" # folder contains test data + format: "nifti" + type: "paired" # one of "paired", "unpaired" or "grouped" + labeled: true + sample_label: "sample" # one of "sample", "all" or None +` + +In the case the `labeled` argument is false, the sample_label is unused, but still must +be passed. Additionally, if the tensors in the files only have one label, irregardles of +the `sample_label` argument, the data loader will only pass the one label to the +network. So, then... Ok, to clarify: * all: for one image, x labels: yields x image label pairs with the same image a. Occurs over all images, in one epoch. * sample: for one image, with x labels: yields 1 image label pair randomly sampled. Occurs for all images in one epoch? — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub<[#559 (comment)](https://github.com/DeepRegNet/DeepReg/pull/559#discussion_r559199902)>, or unsubscribehttps://github.com/notifications/unsubscribe-auth/AATX37OZWLEOLVOIYNVY6JDS2MBQBANCNFSM4UWN4EFA.
> > 
> > 
> > But tests are not related to this docs ticket? @YipengHu
> 
> why not - the config files used for both train and predict - see #634. @mathpluscode please clarify this. Is it true if you want different label_sample during train and predict, then you need to change this field "manually". you can't specify them differently in one config files - this can be addressed later but needs to clarify here.

it's only for true, for inference the sample option is ignored and it will be always "all" as we want to have the results on all labels.

Yes, we need a new issue for this.

NMontanaBrown commented:
> > > > This is fine only thing is that for testing there is no such thing as an epoch. Yipeng
> > > > ________________________________ From: Nina Montana-Brown [notifications@github.com](mailto:notifications@github.com) Sent: Sunday, January 17, 2021 3:51:28 PM To: DeepRegNet/DeepReg [DeepReg@noreply.github.com](mailto:DeepReg@noreply.github.com) Cc: Hu, Yipeng [yipeng.hu@ucl.ac.uk](mailto:yipeng.hu@ucl.ac.uk); Mention [mention@noreply.github.com](mailto:mention@noreply.github.com) Subject: Re: [DeepRegNet/DeepReg] 525: update config file docs (#559) @NMontanaBrown commented on this pull request.
> > > > ________________________________ In docs/source/docs/configuration.md<[#559 (comment)](https://github.com/DeepRegNet/DeepReg/pull/559#discussion_r559199902)>:
> > > > +`yaml +dataset: + dir: + train: "data/test/h5/paired/train1" # folders contains training data + valid: "data/test/h5/paired/valid" # folder contains validation data + test: "data/test/h5/paired/test" # folder contains test data + format: "nifti" + type: "paired" # one of "paired", "unpaired" or "grouped" + labeled: true + sample_label: "sample" # one of "sample", "all" or None +` + +In the case the `labeled` argument is false, the sample_label is unused, but still must +be passed. Additionally, if the tensors in the files only have one label, irregardles of +the `sample_label` argument, the data loader will only pass the one label to the +network. So, then... Ok, to clarify: * all: for one image, x labels: yields x image label pairs with the same image a. Occurs over all images, in one epoch. * sample: for one image, with x labels: yields 1 image label pair randomly sampled. Occurs for all images in one epoch? — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub<[#559 (comment)](https://github.com/DeepRegNet/DeepReg/pull/559#discussion_r559199902)>, or unsubscribehttps://github.com/notifications/unsubscribe-auth/AATX37OZWLEOLVOIYNVY6JDS2MBQBANCNFSM4UWN4EFA.
> > > 
> > > 
> > > But tests are not related to this docs ticket? @YipengHu
> > 
> > 
> > why not - the config files used for both train and predict - see #634. @mathpluscode please clarify this. Is it true if you want different label_sample during train and predict, then you need to change this field "manually". you can't specify them differently in one config files - this can be addressed later but needs to clarify here.
> 
> it's only for true, for inference the sample option is ignored and it will be always "all" as we want to have the results on all labels.
> 
> Yes, we need a new issue for this.

Ok, so after we raise the ticket, is this ready for merge?

YipengHu commented:
> > > > > This is fine only thing is that for testing there is no such thing as an epoch. Yipeng
> > > > > ________________________________ From: Nina Montana-Brown [notifications@github.com](mailto:notifications@github.com) Sent: Sunday, January 17, 2021 3:51:28 PM To: DeepRegNet/DeepReg [DeepReg@noreply.github.com](mailto:DeepReg@noreply.github.com) Cc: Hu, Yipeng [yipeng.hu@ucl.ac.uk](mailto:yipeng.hu@ucl.ac.uk); Mention [mention@noreply.github.com](mailto:mention@noreply.github.com) Subject: Re: [DeepRegNet/DeepReg] 525: update config file docs (#559) @NMontanaBrown commented on this pull request.
> > > > > ________________________________ In docs/source/docs/configuration.md<[#559 (comment)](https://github.com/DeepRegNet/DeepReg/pull/559#discussion_r559199902)>:
> > > > > +`yaml +dataset: + dir: + train: "data/test/h5/paired/train1" # folders contains training data + valid: "data/test/h5/paired/valid" # folder contains validation data + test: "data/test/h5/paired/test" # folder contains test data + format: "nifti" + type: "paired" # one of "paired", "unpaired" or "grouped" + labeled: true + sample_label: "sample" # one of "sample", "all" or None +` + +In the case the `labeled` argument is false, the sample_label is unused, but still must +be passed. Additionally, if the tensors in the files only have one label, irregardles of +the `sample_label` argument, the data loader will only pass the one label to the +network. So, then... Ok, to clarify: * all: for one image, x labels: yields x image label pairs with the same image a. Occurs over all images, in one epoch. * sample: for one image, with x labels: yields 1 image label pair randomly sampled. Occurs for all images in one epoch? — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub<[#559 (comment)](https://github.com/DeepRegNet/DeepReg/pull/559#discussion_r559199902)>, or unsubscribehttps://github.com/notifications/unsubscribe-auth/AATX37OZWLEOLVOIYNVY6JDS2MBQBANCNFSM4UWN4EFA.
> > > > 
> > > > 
> > > > But tests are not related to this docs ticket? @YipengHu
> > > 
> > > 
> > > why not - the config files used for both train and predict - see #634. @mathpluscode please clarify this. Is it true if you want different label_sample during train and predict, then you need to change this field "manually". you can't specify them differently in one config files - this can be addressed later but needs to clarify here.
> > 
> > 
> > it's only for true, for inference the sample option is ignored and it will be always "all" as we want to have the results on all labels.
> > Yes, we need a new issue for this.
> 
> Ok, so after we raise the ticket, is this ready for merge?

if this is true, i don't think we need a new ticket, just clarify saying the label_sample is only for training and we are ok. sorry for misunderstanding from me.

NMontanaBrown commented:
> > > > > > This is fine only thing is that for testing there is no such thing as an epoch. Yipeng
> > > > > > ________________________________ From: Nina Montana-Brown [notifications@github.com](mailto:notifications@github.com) Sent: Sunday, January 17, 2021 3:51:28 PM To: DeepRegNet/DeepReg [DeepReg@noreply.github.com](mailto:DeepReg@noreply.github.com) Cc: Hu, Yipeng [yipeng.hu@ucl.ac.uk](mailto:yipeng.hu@ucl.ac.uk); Mention [mention@noreply.github.com](mailto:mention@noreply.github.com) Subject: Re: [DeepRegNet/DeepReg] 525: update config file docs (#559) @NMontanaBrown commented on this pull request.
> > > > > > ________________________________ In docs/source/docs/configuration.md<[#559 (comment)](https://github.com/DeepRegNet/DeepReg/pull/559#discussion_r559199902)>:
> > > > > > +`yaml +dataset: + dir: + train: "data/test/h5/paired/train1" # folders contains training data + valid: "data/test/h5/paired/valid" # folder contains validation data + test: "data/test/h5/paired/test" # folder contains test data + format: "nifti" + type: "paired" # one of "paired", "unpaired" or "grouped" + labeled: true + sample_label: "sample" # one of "sample", "all" or None +` + +In the case the `labeled` argument is false, the sample_label is unused, but still must +be passed. Additionally, if the tensors in the files only have one label, irregardles of +the `sample_label` argument, the data loader will only pass the one label to the +network. So, then... Ok, to clarify: * all: for one image, x labels: yields x image label pairs with the same image a. Occurs over all images, in one epoch. * sample: for one image, with x labels: yields 1 image label pair randomly sampled. Occurs for all images in one epoch? — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub<[#559 (comment)](https://github.com/DeepRegNet/DeepReg/pull/559#discussion_r559199902)>, or unsubscribehttps://github.com/notifications/unsubscribe-auth/AATX37OZWLEOLVOIYNVY6JDS2MBQBANCNFSM4UWN4EFA.
> > > > > 
> > > > > 
> > > > > But tests are not related to this docs ticket? @YipengHu
> > > > 
> > > > 
> > > > why not - the config files used for both train and predict - see #634. @mathpluscode please clarify this. Is it true if you want different label_sample during train and predict, then you need to change this field "manually". you can't specify them differently in one config files - this can be addressed later but needs to clarify here.
> > > 
> > > 
> > > it's only for true, for inference the sample option is ignored and it will be always "all" as we want to have the results on all labels.
> > > Yes, we need a new issue for this.
> > 
> > 
> > Ok, so after we raise the ticket, is this ready for merge?
> 
> if this is true, i don't think we need a new ticket, just clarify saying the label_sample is only for training and we are ok. sorry for misunderstanding from me.

@YipengHu  Just to confirm, @mathpluscode is correct - in test mode we use "all", see the code here: https://github.com/DeepRegNet/DeepReg/blob/859df2a91564d16c43ea761ad272de568b8784bb/deepreg/dataset/load.py#L27 

So I will add this comment to the docs.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/DeepRegNet/DeepReg/inspections/5cd42468-9991-479e-a736-b44d126db935) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/DeepRegNet/DeepReg/inspections/5cd42468-9991-479e-a736-b44d126db935) completed: **No new issues**

NMontanaBrown commented:
@mathpluscode @YipengHu Thanks both, ticket finally closing! 🎆 🍾
